src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -0,0 +1,403 @@
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool ShouldRescanOccur(Interop.CoreServices.FSEventStreamEventFlags flags)
+        {
+            // Check if the MustScanSubDirs flag exists
+            bool shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagMustScanSubDirs) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagMustScanSubDirs);
+            
+            // Check if the UserDropped flag exists; meaning the UserMode daemon dropped a change coming from the kernel
+            if (shouldRescan == false)
+            {
+                shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagUserDropped) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagUserDropped);
+            }
+
+            // Check if the KernelDropped flag exists; meaning the Kernel dropped a flag before telling the UserMode daemon about it
+            if (shouldRescan == false)
+            {
+                shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagKernelDropped) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagKernelDropped);
+            }
+
+            // Check if the RootChanged flag exists; meaning the Root folder changed in some way that could affect all child items and requires a rescan to rebuild an accurate snapshot.
+            if (shouldRescan == false)
+            {
+                shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagRootChanged) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagRootChanged);
+            }
+
+            // Check if the Mount flag exists; meaning a whole directory has been added under the root, which would require a rescan to rebuild an accurate snapshot.
+            if (shouldRescan == false)
+            {
+                shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagMount) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagMount);
+            }
+
+            // Check if the Unmount flag exists; meaning a whole directory has been removed from under the root, which would require a rescan to rebuild an accurate snapshot
+            if (shouldRescan == false)
+            {
+                shouldRescan = ((flags & Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagUnmount) == Interop.CoreServices.FSEventStreamEventFlags.kFSEventStreamEventFlagUnmount);
+            }
+
+            return shouldRescan;
+        }
+
+        private bool DoesPathPassNameFilter(string eventPath)
+        {
+            bool doesPathPass = true;
+            
+            // If we shouldn't include subdirectories, check if this path's parent is the watch directory
+            if (_includeSubdirectories == false)
```

Comment:
There's no way via the underlying API to tell it to not recur into subdirectories?  Too bad.